test/distributed/_composable/fsdp/test_fully_shard_training.py

```diff
@@ -710,13 +713,16 @@ class TestFullyShard2DTraining(FSDPTest):
     def world_size(self) -> int:
         return min(4, torch.cuda.device_count())
 
-    @skip_if_lt_x_gpu(2)
-    def test_train_parity_2d_mlp(self):
+    def init_global_mesh(self) -> DeviceMesh:
         # Prefer to test with >=4 GPUs, but for 2 GPUs, use 2-way TP
         dp_size = 2 if self.world_size > 2 else 1
-        global_mesh = init_device_mesh(
+        return init_device_mesh(
             "cuda", (dp_size, self.world_size // dp_size), mesh_dim_names=("dp", "tp")
         )
+
+    @skip_if_lt_x_gpu(2)
+    def test_train_parity_2d_mlp(self):
```

Comment:
Should we make optimizer an option in .run_subtests? 
Actually, I can try this after you land, since I have observed parameters are off by a little bit more than what torch.allclose() tolerates by default in the unit tests.

Context: https://github.com/pytorch/pytorch/pull/121774
Local test is fine but CI is failing.

Reply:
I think since Adam is fine/passing, maybe we can leave the test always testing optimizer here since we want to check the end-to-end resume flow? Then, you can investigate the AdamW case separately?